File: Android/app/src/main/java/alexa4/friendphoto/repositories/DataRepository.java
package alexa4.friendphoto.repositories;

import android.content.Context;
import android.util.Log;

import java.util.ArrayList;

import alexa4.friendphoto.controllers.DbController;
import alexa4.friendphoto.controllers.FriendsController;
import alexa4.friendphoto.controllers.PhotosController;
import alexa4.friendphoto.models.Friend;
import alexa4.friendphoto.models.Photo;
import alexa4.friendphoto.models.User;
import alexa4.friendphoto.ui.ActivityCallback;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;


/**
 * Repository that contains both friends and photos controllers,
 * this repository makes network calls and db calls
 */
public class DataRepository implements AuthCallback {
    private static final String TAG = "DATA_REPOSITORY";
    // Network client
    private final OkHttpClient client = new OkHttpClient();
    // Data controllers
    private final FriendsController mFriends = new FriendsController();
    private final PhotosController mPhotos = new PhotosController();
    private User mUser;

    private DbController mDbController;
    private StorageProvider mStorage = new StorageProvider();

    // The link to UI
    private ActivityCallback mUI;

    public DataRepository(ActivityCallback UI) {
        mUI = UI;
        mDbController = new DbController(mUI.getContext());
        mStorage.initialize(mUI.getContext()
                .getSharedPreferences("DataRepository", Context.MODE_PRIVATE));
    }

    /**
     * RX call of downloading photos of specified friend
     *
     * @param friendId id of friend whose photos need to download
     */
    public Observable<ArrayList<Photo>> downloadPhotos(int friendId) {
        return Observable.<ArrayList<Photo>>create(emitter -> {
            Request request = new Request.Builder()
                    .url("https://api.vk.com/method/photos.get?owner_id="
                            + friendId + "&album_id=wall&v=5.95&access_token="
                            + mUser.mToken)
                    .build();
            try (Response response = client.newCall(request).execute()) {
                if (!mPhotos.isIdenticalId(friendId) && response.isSuccessful()) {
                    mPhotos.setOwnerId(friendId);
                    mPhotos.setPhotos(response.body().string());
                }
            }
            emitter.onNext(mPhotos.mPhotos);
            emitter.onComplete();
        })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()
                );
    }

    /**
     * Return observable which should download information about friends
     */
    public Observable<ArrayList<Friend>> downloadFriends() {
        return Observable.<ArrayList<Friend>>create(emitter -> {
            // Read from DB if there is first launch
            if (!mFriends.isLoaded())
                mFriends.setFromDb(mDbController.readFriends());

            // Download from network if there is no data after DB reading
            if (!mFriends.isLoaded()) {

                Request request = new Request.Builder()
                        .url("https://api.vk.com/method/friends.get?fields=photo_100, status&v=5.95&access_token="
                                + mUser.mToken)
                        .build();
                try (Response response = client.newCall(request).execute()) {
                    if (response.isSuccessful()) {
                        mFriends.setFriends(response.body().string());
                        mDbController.saveFriends(mFriends.mFriends);
                    }
                }

            }
            emitter.onNext(mFriends.mFriends);
            emitter.onComplete();
        })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()
                );
    }


    /**
     * Dispose all subscription and clear data
     */
    public void dispose() {
        mFriends.dispose();
        mPhotos.dispose();
    }


    /**
     * Get data of user authentication process and pull result
     * to UI through callback
     *
     * @param fragments the result string of authentication process
     */
    @Override
    public void sendUserAuthFragment(String fragments) {
        Log.d(TAG, fragments);
        String[] data = fragments.split("&");

        // If user cancel authentication
        if (data[0].contains("error")) {
            mUI.authenticateUser(false);
            return;
        }

        // Initialize user data
        mUser = new User(data[0].replace("access_token=", ""),
                Integer.parseInt(data[2].replace("user_id=", "")));
        mStorage.saveUserData(mUser); //save data to prefs
        mUI.authenticateUser(true); //notify UI
    }

    /**
     * Check is user authenticated and get its info if true
     * This method must be called before any transactions
     *
     * @return is user authenticated
     */
    public boolean checkAuthentication() {
        if (mStorage.isUserAuthenticated()) {
            try {
                mUser = mStorage.getUserData();

                return true;
            } catch (Exception e) {
                return false;
            }
        } else return false;
    }
}

File: Flutter/friend_photo_flutter/lib/src/repositories/data_repository.dart
import 'package:friend_photo_flutter/src/controllers/db_controller.dart';
import 'package:friend_photo_flutter/src/controllers/friends_controller.dart';
import 'package:friend_photo_flutter/src/controllers/photos_controller.dart';
import 'package:friend_photo_flutter/src/models/friend.dart';
import 'package:friend_photo_flutter/src/models/photo.dart';
import 'package:friend_photo_flutter/src/models/user.dart';
import 'package:friend_photo_flutter/src/repositories/storage_repo.dart';
import 'package:http/http.dart' as http;

/// Repository that provide full data about friends, photos and user
class DataRepository {
  final FriendsController _friends = FriendsController();
  final PhotosController _photos = PhotosController();
  final StorageRepository _storage = StorageRepository();
  final DbController _db = DbController();
  User _user;

  /// Check is user authenticated
  /// If true, then initialize [_user] with data
  Future<bool> checkAuthentication() async {
    if (await _storage.isUserAuthenticated()) {
      _user = await _storage.getUserData();

      // Read friends list
      var list = await _db.getFriendsList();
      _friends.setFromDb(list);

      return true;
    }
    return false;
  }

  /// Get list of friends
  /// If they were in DB then return momentally
  /// else download them through API
  Future<List<Friend>> getFriendsList() async {
    if (_friends.friends.length != 0) return _friends.friends;

    var response =
        await http.get('https://api.vk.com/method/friends.get?fields=photo_100,'
            'status&v=5.95&access_token=${_user.accessToken}');
    if (response.statusCode == 200) {
      _friends.setFriends(response.body);

      // Save friends to DB
      await _db.setFriendsList(_friends.friends);

      return _friends.friends;
    } else
      throw Exception();
  }

  /// Get list of photos by specified [friendId]
  /// If that photos alredy downloaded, then return them
  Future<List<Photo>> getPhotosList(int friendId) async {
    if (_photos.isIdential(friendId)) return _photos.photos;

    var response =
        await http.get('https://api.vk.com/method/photos.get?owner_id='
            '$friendId&album_id=wall&v=5.95&access_token=${_user.accessToken}');

    // If everything is OK, then return new list
    if (response.statusCode == 200) {
      _photos.ownerId = friendId;
      _photos.setPhotos(response.body);
      return _photos.photos;
    } else
      throw Exception();
  }

  /// Authenticate user from input [url]
  /// There can be error if authentication attempt is wrong
  /// or can be access_token and user_id if everything is ok
  bool authenticateUser(String url) {
    print(url);
    List<String> data = url.split('&');
    // If this is wrong authentication
    if (data[0].contains('error')) return false;

    _user = new User(
        accessToken: data[0].replaceAll('access_token=', ''),
        id: int.parse(data[2].replaceAll('user_id=', '')));
    _storage.saveUserData(_user);

    return true;
  }

  /// Free memory
  void dispose() {
    _friends.dispose();
    _photos.dispose();
  }
}
